AsyncDisplayKit/Details/ASDataController+Subclasses.h

```diff
@@ -0,0 +1,159 @@
+
+/**
+ * Provides a collection of index paths for nodes of the given kind that are currently in the editing store
+ */
+- (NSArray *)indexPathsForEditingNodesOfKind:(NSString *)kind;
+
+/**
+ * Read-only access to the underlying editing nodes of the given kind
+ */
+- (NSMutableArray *)editingNodesOfKind:(NSString *)kind;
+
+/**
+ * Read only access to the underlying completed nodes of the given kind
+ */
+- (NSMutableArray *)completedNodesOfKind:(NSString *)kind;
+
+#pragma mark - Node sizing
+
+/**
+ * Measure and layout the given nodes in optimized batches, constraining each to a given size in `constrainedSizeForNodeOfKind:atIndexPath:`.
+ */
+- (void)batchLayoutNodes:(NSArray *)nodes ofKind:(NSString *)kind atIndexPaths:(NSArray *)indexPaths completion:(void (^)(NSArray *nodes, NSArray *indexPaths))completionBlock;
+
+/*
+ * Perform measurement and layout of loaded nodes on the main thread, skipping unloaded nodes.
+ *
+ * @discussion Once nodes have loaded their views, we can't layout in the background so this is a chance
+ * to do so immediately on the main thread.
+ */
+- (void)layoutLoadedNodes:(NSArray *)nodes ofKind:(NSString *)kind atIndexPaths:(NSArray *)indexPaths;
+
+/**
+ * Provides the size range for a specific node during the layout process.
+ */
+- (ASSizeRange)constrainedSizeForNodeOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath;
+
+#pragma mark - Node & Section Insertion/Deletion API
+
+/**
+ * Inserts the given nodes of the specified kind into the backing store, calling completion on the main thread when the write finishes.
+ */
+- (void)insertNodes:(NSArray *)nodes ofKind:(NSString *)kind atIndexPaths:(NSArray *)indexPaths completion:(void (^)(NSArray *nodes, NSArray *indexPaths))completionBlock;
+
+/**
+ * Deletes the given nodes of the specified kind in the backing store, calling completion on the main thread when the deletion finishes.
+ */
+- (void)deleteNodesOfKind:(NSString *)kind atIndexPaths:(NSArray *)indexPaths completion:(void (^)(NSArray *nodes, NSArray *indexPaths))completionBlock;
+
+/**
+ * Inserts the given sections of the specified kind in the backing store, calling completion on the main thread when finished.
+ */
+- (void)insertSections:(NSMutableArray *)sections ofKind:(NSString *)kind atIndexSet:(NSIndexSet *)indexSet completion:(void (^)(NSArray *sections, NSIndexSet *indexSet))completionBlock;
+
+/**
+ * Deletes the given sections of the specified kind in the backing store, calling completion on the main thread when finished.
+ */
+- (void)deleteSectionsOfKind:(NSString *)kind atIndexSet:(NSIndexSet *)indexSet completion:(void (^)(NSIndexSet *indexSet))completionBlock;
+
+#pragma mark - Data Manipulation Hooks
+
+/**
+ * Notifies the subclass to perform any work needed before the data controller is reloaded entirely
+ *
+ * @discussion This method will be performed before the data controller enters its editing queue, usually on the main
```

Comment:
Is this /usually/ on main, or always?  If it is not a guarantee of the API, can we eliminate this method and have only -willReloadData?